base/shell/explorer/syspager.cpp

```diff
@@ -1552,6 +1596,164 @@ LRESULT CSysPagerWnd::OnCopyData(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL&
+            if ((pIcon->IconData.dwState & NIS_HIDDEN) == NIS_HIDDEN)
+            {
+                pIcon->IconData.dwState &= ~NIS_HIDDEN;
+                tbbi.fsState &= ~TBSTATE_HIDDEN;
+                obj->Toolbar.IncVisibleButtons();
+            }
+        }
+        else if (pIcon->dwPreference != ALWAYS_SHOW )
+        {
+            if ((pIcon->IconData.dwState & NIS_HIDDEN) != NIS_HIDDEN)
+            {
+                pIcon->IconData.dwState |= NIS_HIDDEN;
+                tbbi.fsState |= TBSTATE_HIDDEN;
+                obj->Toolbar.DecVisibleButtons();
+            }
+        }
+    }
+    else if (pIcon->dwPreference != ALWAYS_HIDE)
+    {
+        if ((pIcon->IconData.dwState & NIS_HIDDEN) == NIS_HIDDEN)
+         {
+             pIcon->IconData.dwState &= ~NIS_HIDDEN;
+             tbbi.fsState &= ~TBSTATE_HIDDEN;
+             obj->Toolbar.IncVisibleButtons();
+         }
+    }
+    obj->Toolbar.SetButtonInfo(index, &tbbi);
+
+    return TRUE;
+}
+
+LRESULT CSysPagerWnd::OnMouseLeave(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (bExpanded)
+    {
+        AutoTray(FALSE);
+        bExpanded = FALSE;
+    }
+    return 0;
+}
+
+VOID CSysPagerWnd::AutoTray(BOOL expand)
+{
+
+    PullServicesEnabled();
+    EnumIcons(EnumProc, reinterpret_cast<LPARAM>(GetIconList()), expand);
+    bExpanded = expand;
+
+    HINSTANCE hInst = reinterpret_cast<HINSTANCE>(GetModuleHandle(NULL));
+ 
+    NOTIFYICONDATA iData;
+    iData.cbSize = sizeof(iData);
+    iData.hWnd = m_hWnd;
+    iData.uID = IDI_TRAY_EXPAND;
+    iData.uFlags = NIF_ICON | NIF_MESSAGE;
+    iData.uCallbackMessage = TNWM_AUTOTRAY;
+    if (expand)
+    {
+        iData.hIcon = LoadIcon(hInst, MAKEINTRESOURCE(IDI_TRAY_COLAPSE));
+    }
+    else
+    {
+        iData.hIcon = LoadIcon(hInst, MAKEINTRESOURCE(IDI_TRAY_EXPAND));
+    }
+    NotifyIcon(NIM_MODIFY, &iData);
+
+    NMHDR nmh = {GetParent(), 0, NTNWM_REALIGN};
+    GetParent().SendMessage(WM_NOTIFY, 0, (LPARAM) &nmh);
+}
+
+LRESULT CSysPagerWnd::OnTrayExpand(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    switch (lParam)
+    {
+        case WM_LBUTTONUP:
+        {
+            AutoTray(TRUE);
+            break;
+        }
+    }
+    return S_OK;
+}
+
+LRESULT CSysPagerWnd::OnEnableAutoTray(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    HINSTANCE hInst = (HINSTANCE)GetModuleHandle(NULL);
+ 
+    NOTIFYICONDATA iData;
+    iData.cbSize = sizeof(iData);
+    iData.hWnd = m_hWnd;
+    iData.uID = IDI_TRAY_EXPAND;
+    iData.uFlags = NIF_ICON | NIF_MESSAGE;
+    iData.uCallbackMessage = TNWM_AUTOTRAY;
+    iData.hIcon = LoadIcon(hInst, MAKEINTRESOURCE(IDI_TRAY_EXPAND));
+    if (wParam)
+    {
+        NotifyIcon(NIM_ADD, &iData);
+    }
+    else
+    {
+        NotifyIcon(NIM_DELETE, &iData);
+        return S_OK;
+    }
+  //  INT ButtonCount = Toolbar.GetVisibleButtonCount() - 1;
+  //  Toolbar.MoveButton(ButtonCount, 0);
+    AutoTray(FALSE);
```

Comment:
If you move the `AutoTray` into the `if (wParam)` then you don't need the `return S_OK;` in the `else